pkg/xds/experimental/client/listener.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// listener is a helper structure for watchers focused on handling a single, registered callback.
+type listener struct {
+	ID      uint64
+	typeUrl string
+	cb      WatcherCallback
+	trigger chan uint64
+
+	log       *slog.Logger
+	resources xds.ResourceSource
+
+	lastVersion uint64
+	mux         lock.Mutex
+	nextVersion uint64
+}
+
+var _ xds.ResourceVersionObserver = (*listener)(nil)
+
+// HandleNewResourceVersion implements xds.ResourceVersionObserver.
+// It triggers the callback process when a new version appears.
+func (l *listener) HandleNewResourceVersion(typeUrl string, version uint64) {
+	if typeUrl != l.typeUrl {
+		return
+	}
+	l.mux.Lock()
+	defer l.mux.Unlock()
+	if version <= l.lastVersion {
+		return
+	}
+	l.nextVersion = version
+	select {
+	case l.trigger <- version:
+	default:
+	}
+}
+
+// process waits for the listener trigger (new resource version) and invokes the callback function.
+// It needs to be done asynchronyously from HandleNewResourceVersion, because
+// the cache invoking the function holds a lock in the resources.
+func (l *listener) process() {
+	for ver := range l.trigger {
+		l.mux.Lock()
+		version := l.nextVersion
+		if ver != version {
+			l.log.Debug("skipped versions", "last-processed-version", l.lastVersion, "next-processed-version", l.nextVersion, "trigger-version", ver)
```

Comment:
`l.lastVersion` equals `version` you can Unlock before line 54 and pass version to the log entry

Reply:
This comment is made to outdated code. Please review the latest version